src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs

```diff
@@ -54,24 +54,34 @@ internal HttpParser(bool showErrorDetails, bool disableHttp1LineFeedTerminators)
     private const byte ByteQuestionMark = (byte)'?';
     private const byte BytePercentage = (byte)'%';
     private const int MinTlsRequestSize = 1; // We need at least 1 byte to check for a proper TLS request line
+    private static ReadOnlySpan<byte> RequestLineDelimiters => [ByteLF, 0];
 
     /// <summary>
     /// This API supports framework infrastructure and is not intended to be used
     /// directly from application code.
     /// </summary>
     public bool ParseRequestLine(TRequestHandler handler, ref SequenceReader<byte> reader)
     {
-        if (reader.TryReadTo(out ReadOnlySpan<byte> requestLine, ByteLF, advancePastDelimiter: true))
+        // Find the next delimiter.
+        if (!reader.TryReadToAny(out ReadOnlySpan<byte> requestLine, RequestLineDelimiters, advancePastDelimiter: false))
         {
-            ParseRequestLine(handler, requestLine);
-            return true;
+            return false;
         }
 
-        return false;
-    }
+        // Consume the delimiter.
+        var foundDelimiter = reader.TryRead(out var next);
+        Debug.Assert(foundDelimiter);
+        // If null character found, or request line is empty
+        if (next == 0 || requestLine.Length == 0)
+        {
+            // Rewind and re-read to format error message correctly
+            reader.Rewind(requestLine.Length + 1);
+            var readResult = reader.TryReadExact(requestLine.Length + 1, out var requestLineSequence);
+            Debug.Assert(readResult);
+            requestLine = requestLineSequence.IsSingleSegment ? requestLineSequence.FirstSpan : requestLineSequence.ToArray();
```

Comment:
Is this going to include the LF for an empty line?

Reply:
Yeah it would in theory is that a problem?

> | Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException: Invalid request line: '\x0A'

I haven't seen a scenario where we'll actually get an empty request line since we have an early check for an empty request:
https://github.com/dotnet/aspnetcore/blob/d96d2728dff8d1ded8c5b33da9921cdcb6c6d94e/src/Servers/Kestrel/Core/src/Internal/Http/Http1Connection.cs#L173-L177
So it's mostly defensive coding.

Reply:
No, I just didn't know if trimming would be preferable.  If we're escaping whitespace in the error message, then including it seems sensible.